// @pulumi/aws/cloudtrail/trail.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Trail = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a CloudTrail resource.
 *
 * > **Tip:** For a multi-region trail, this resource must be in the home region of the trail.
 *
 * > **Tip:** For an organization trail, this resource must be in the master account of the organization.
 *
 * ## Example Usage
 * ### Basic
 *
 * Enable CloudTrail to capture all compatible management events in region.
 * For capturing events from services like IAM, `includeGlobalServiceEvents` must be enabled.
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const current = aws.getCallerIdentity({});
 * const bucket = new aws.s3.Bucket("bucket", {});
 * const bucketPolicy = new aws.s3.BucketPolicy("bucketPolicy", {
 *     bucket: bucket.id,
 *     policy: pulumi.all([bucket.id, bucket.id, current]).apply(([bucketId, bucketId1, current]) => `  {
 *       "Version": "2012-10-17",
 *       "Statement": [
 *           {
 *               "Sid": "AWSCloudTrailAclCheck",
 *               "Effect": "Allow",
 *               "Principal": {
 *                 "Service": "cloudtrail.amazonaws.com"
 *               },
 *               "Action": "s3:GetBucketAcl",
 *               "Resource": "arn:aws:s3:::${bucketId}"
 *           },
 *           {
 *               "Sid": "AWSCloudTrailWrite",
 *               "Effect": "Allow",
 *               "Principal": {
 *                 "Service": "cloudtrail.amazonaws.com"
 *               },
 *               "Action": "s3:PutObject",
 *               "Resource": "arn:aws:s3:::${bucketId1}/prefix/AWSLogs/${current.accountId}/*",
 *               "Condition": {
 *                   "StringEquals": {
 *                       "s3:x-amz-acl": "bucket-owner-full-control"
 *                   }
 *               }
 *           }
 *       ]
 *   }
 * `),
 * });
 * const foobar = new aws.cloudtrail.Trail("foobar", {
 *     s3BucketName: bucket.id,
 *     s3KeyPrefix: "prefix",
 *     includeGlobalServiceEvents: false,
 * });
 * ```
 * ### Data Event Logging
 *
 * CloudTrail can log [Data Events](https://docs.aws.amazon.com/awscloudtrail/latest/userguide/logging-data-events-with-cloudtrail.html) for certain services such as S3 bucket objects and Lambda function invocations. Additional information about data event configuration can be found in the following links:
 *
 * * [CloudTrail API DataResource documentation](https://docs.aws.amazon.com/awscloudtrail/latest/APIReference/API_DataResource.html) (for basic event selector).
 * * [CloudTrail API AdvancedFieldSelector documentation](https://docs.aws.amazon.com/awscloudtrail/latest/APIReference/API_AdvancedFieldSelector.html) (for advanced event selector).
 * ### Logging All Lambda Function Invocations By Using Basic Event Selectors
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const bucket = new aws.s3.Bucket("bucket", {});
 * const example = new aws.cloudtrail.Trail("example", {
 *     s3BucketName: bucket.id,
 *     s3KeyPrefix: "prefix",
 *     eventSelectors: [{
 *         readWriteType: "All",
 *         includeManagementEvents: true,
 *         dataResources: [{
 *             type: "AWS::Lambda::Function",
 *             values: ["arn:aws:lambda"],
 *         }],
 *     }],
 * });
 * ```
 * ### Logging All S3 Bucket Object Events By Using Basic Event Selectors
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const bucket = new aws.s3.Bucket("bucket", {});
 * const example = new aws.cloudtrail.Trail("example", {
 *     s3BucketName: bucket.id,
 *     s3KeyPrefix: "prefix",
 *     eventSelectors: [{
 *         readWriteType: "All",
 *         includeManagementEvents: true,
 *         dataResources: [{
 *             type: "AWS::S3::Object",
 *             values: ["arn:aws:s3:::"],
 *         }],
 *     }],
 * });
 * ```
 * ### Logging Individual S3 Bucket Events By Using Basic Event Selectors
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const important-bucket = aws.s3.getBucket({
 *     bucket: "important-bucket",
 * });
 * const example = new aws.cloudtrail.Trail("example", {
 *     s3BucketName: important_bucket.then(important_bucket => important_bucket.id),
 *     s3KeyPrefix: "prefix",
 *     eventSelectors: [{
 *         readWriteType: "All",
 *         includeManagementEvents: true,
 *         dataResources: [{
 *             type: "AWS::S3::Object",
 *             values: [important_bucket.then(important_bucket => `${important_bucket.arn}/`)],
 *         }],
 *     }],
 * });
 * ```
 * ### Logging All S3 Bucket Object Events Except For Two S3 Buckets By Using Advanced Event Selectors
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const not_important_bucket_1 = pulumi.output(aws.s3.getBucket({
 *     bucket: "not-important-bucket-1",
 * }));
 * const not_important_bucket_2 = pulumi.output(aws.s3.getBucket({
 *     bucket: "not-important-bucket-2",
 * }));
 * const example = new aws.cloudtrail.Trail("example", {
 *     advancedEventSelectors: [
 *         {
 *             fieldSelectors: [
 *                 {
 *                     equals: ["Data"],
 *                     field: "eventCategory",
 *                 },
 *                 {
 *                     field: "resources.ARN",
 *                     notEquals: [
 *                         pulumi.interpolate`${not_important_bucket_1.arn}/`,
 *                         pulumi.interpolate`${not_important_bucket_2.arn}/`,
 *                     ],
 *                 },
 *                 {
 *                     equals: ["AWS::S3::Object"],
 *                     field: "resources.type",
 *                 },
 *             ],
 *             name: "Log all S3 buckets objects events except for two S3 buckets",
 *         },
 *         {
 *             fieldSelectors: [{
 *                 equals: ["Management"],
 *                 field: "eventCategory",
 *             }],
 *             name: "Log readOnly and writeOnly management events",
 *         },
 *     ],
 * });
 * ```
 * ### Logging Individual S3 Buckets And Specific Event Names By Using Advanced Event Selectors
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const important_bucket_1 = pulumi.output(aws.s3.getBucket({
 *     bucket: "important-bucket-1",
 * }));
 * const important_bucket_2 = pulumi.output(aws.s3.getBucket({
 *     bucket: "important-bucket-2",
 * }));
 * const important_bucket_3 = pulumi.output(aws.s3.getBucket({
 *     bucket: "important-bucket-3",
 * }));
 * const example = new aws.cloudtrail.Trail("example", {
 *     advancedEventSelectors: [
 *         {
 *             fieldSelectors: [
 *                 {
 *                     equals: ["Data"],
 *                     field: "eventCategory",
 *                 },
 *                 {
 *                     equals: [
 *                         "PutObject",
 *                         "DeleteObject",
 *                     ],
 *                     field: "eventName",
 *                 },
 *                 {
 *                     //The trailing slash is intentional; do not exclude it.
 *                     equals: [
 *                         pulumi.interpolate`${important_bucket_1.arn}/`,
 *                         pulumi.interpolate`${important_bucket_2.arn}/`,
 *                     ],
 *                     field: "resources.ARN",
 *                 },
 *                 {
 *                     equals: ["false"],
 *                     field: "readOnly",
 *                 },
 *                 {
 *                     equals: ["AWS::S3::Object"],
 *                     field: "resources.type",
 *                 },
 *             ],
 *             name: "Log PutObject and DeleteObject events for two S3 buckets",
 *         },
 *         {
 *             fieldSelectors: [
 *                 {
 *                     equals: ["Data"],
 *                     field: "eventCategory",
 *                 },
 *                 {
 *                     field: "eventName",
 *                     startsWiths: ["Delete"],
 *                 },
 *                 {
 *                     //The trailing slash is intentional; do not exclude it.
 *                     equals: [pulumi.interpolate`${important_bucket_3.arn}/important-prefix`],
 *                     field: "resources.ARN",
 *                 },
 *                 {
 *                     equals: ["false"],
 *                     field: "readOnly",
 *                 },
 *                 {
 *                     equals: ["AWS::S3::Object"],
 *                     field: "resources.type",
 *                 },
 *             ],
 *             name: "Log Delete* events for one S3 bucket",
 *         },
 *     ],
 * });
 * ```
 * ### Sending Events to CloudWatch Logs
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const current = aws.getPartition({});
 * const exampleLogGroup = new aws.cloudwatch.LogGroup("exampleLogGroup", {});
 * const testRole = new aws.iam.Role("testRole", {assumeRolePolicy: current.then(current => `{
 *   "Version": "2012-10-17",
 *   "Statement": [
 *     {
 *       "Sid": "",
 *       "Effect": "Allow",
 *       "Principal": {
 *         "Service": "cloudtrail.${current.dnsSuffix}"
 *       },
 *       "Action": "sts:AssumeRole"
 *     }
 *   ]
 * }
 * `)});
 * const testRolePolicy = new aws.iam.RolePolicy("testRolePolicy", {
 *     role: testRole.id,
 *     policy: `{
 *   "Version": "2012-10-17",
 *   "Statement": [
 *     {
 *       "Sid": "AWSCloudTrailCreateLogStream",
 *       "Effect": "Allow",
 *       "Action": [
 *         "logs:CreateLogStream",
 *         "logs:PutLogEvents"
 *       ],
 *       "Resource": "${aws_cloudwatch_log_group.test.arn}:*"
 *     }
 *   ]
 * }
 * `,
 * });
 * const bucket = new aws.s3.Bucket("bucket", {});
 * const exampleTrail = new aws.cloudtrail.Trail("exampleTrail", {
 *     s3BucketName: data.aws_s3_bucket["important-bucket"].id,
 *     s3KeyPrefix: "prefix",
 *     cloudWatchLogsRoleArn: testRole.arn,
 *     cloudWatchLogsGroupArn: pulumi.interpolate`${exampleLogGroup.arn}:*`,
 * });
 * // CloudTrail requires the Log Stream wildcard
 * ```
 *
 * ## Import
 *
 * Cloudtrails can be imported using the `name`, e.g.,
 *
 * ```sh
 *  $ pulumi import aws:cloudtrail/trail:Trail sample my-sample-trail
 * ```
 */
class Trail extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["advancedEventSelectors"] = state ? state.advancedEventSelectors : undefined;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["cloudWatchLogsGroupArn"] = state ? state.cloudWatchLogsGroupArn : undefined;
            inputs["cloudWatchLogsRoleArn"] = state ? state.cloudWatchLogsRoleArn : undefined;
            inputs["enableLogFileValidation"] = state ? state.enableLogFileValidation : undefined;
            inputs["enableLogging"] = state ? state.enableLogging : undefined;
            inputs["eventSelectors"] = state ? state.eventSelectors : undefined;
            inputs["homeRegion"] = state ? state.homeRegion : undefined;
            inputs["includeGlobalServiceEvents"] = state ? state.includeGlobalServiceEvents : undefined;
            inputs["insightSelectors"] = state ? state.insightSelectors : undefined;
            inputs["isMultiRegionTrail"] = state ? state.isMultiRegionTrail : undefined;
            inputs["isOrganizationTrail"] = state ? state.isOrganizationTrail : undefined;
            inputs["kmsKeyId"] = state ? state.kmsKeyId : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["s3BucketName"] = state ? state.s3BucketName : undefined;
            inputs["s3KeyPrefix"] = state ? state.s3KeyPrefix : undefined;
            inputs["snsTopicName"] = state ? state.snsTopicName : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["tagsAll"] = state ? state.tagsAll : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.s3BucketName === undefined) && !opts.urn) {
                throw new Error("Missing required property 's3BucketName'");
            }
            inputs["advancedEventSelectors"] = args ? args.advancedEventSelectors : undefined;
            inputs["cloudWatchLogsGroupArn"] = args ? args.cloudWatchLogsGroupArn : undefined;
            inputs["cloudWatchLogsRoleArn"] = args ? args.cloudWatchLogsRoleArn : undefined;
            inputs["enableLogFileValidation"] = args ? args.enableLogFileValidation : undefined;
            inputs["enableLogging"] = args ? args.enableLogging : undefined;
            inputs["eventSelectors"] = args ? args.eventSelectors : undefined;
            inputs["includeGlobalServiceEvents"] = args ? args.includeGlobalServiceEvents : undefined;
            inputs["insightSelectors"] = args ? args.insightSelectors : undefined;
            inputs["isMultiRegionTrail"] = args ? args.isMultiRegionTrail : undefined;
            inputs["isOrganizationTrail"] = args ? args.isOrganizationTrail : undefined;
            inputs["kmsKeyId"] = args ? args.kmsKeyId : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["s3BucketName"] = args ? args.s3BucketName : undefined;
            inputs["s3KeyPrefix"] = args ? args.s3KeyPrefix : undefined;
            inputs["snsTopicName"] = args ? args.snsTopicName : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["homeRegion"] = undefined /*out*/;
            inputs["tagsAll"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(Trail.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Trail resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Trail(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Trail.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Trail.__pulumiType;
    }
}
exports.Trail = Trail;
/** @internal */
Trail.__pulumiType = 'aws:cloudtrail/trail:Trail';
