// @pulumi/aws/appstream/fleet.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Fleet = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides an AppStream fleet.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const testFleet = new aws.appstream.Fleet("test_fleet", {
 *     computeCapacity: {
 *         desiredInstances: 1,
 *     },
 *     description: "test fleet",
 *     displayName: "test-fleet",
 *     enableDefaultInternetAccess: false,
 *     fleetType: "ON_DEMAND",
 *     idleDisconnectTimeoutInSeconds: 15,
 *     imageName: "Amazon-AppStream2-Sample-Image-02-04-2019",
 *     instanceType: "stream.standard.large",
 *     maxUserDurationInSeconds: 600,
 *     tags: {
 *         TagName: "tag-value",
 *     },
 *     vpcConfig: {
 *         subnetIds: ["subnet-06e9b13400c225127"],
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * `aws_appstream_fleet` can be imported using the id, e.g.,
 *
 * ```sh
 *  $ pulumi import aws:appstream/fleet:Fleet example fleetNameExample
 * ```
 */
class Fleet extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["computeCapacity"] = state ? state.computeCapacity : undefined;
            inputs["createdTime"] = state ? state.createdTime : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["disconnectTimeoutInSeconds"] = state ? state.disconnectTimeoutInSeconds : undefined;
            inputs["displayName"] = state ? state.displayName : undefined;
            inputs["domainJoinInfo"] = state ? state.domainJoinInfo : undefined;
            inputs["enableDefaultInternetAccess"] = state ? state.enableDefaultInternetAccess : undefined;
            inputs["fleetType"] = state ? state.fleetType : undefined;
            inputs["iamRoleArn"] = state ? state.iamRoleArn : undefined;
            inputs["idleDisconnectTimeoutInSeconds"] = state ? state.idleDisconnectTimeoutInSeconds : undefined;
            inputs["imageArn"] = state ? state.imageArn : undefined;
            inputs["imageName"] = state ? state.imageName : undefined;
            inputs["instanceType"] = state ? state.instanceType : undefined;
            inputs["maxUserDurationInSeconds"] = state ? state.maxUserDurationInSeconds : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["state"] = state ? state.state : undefined;
            inputs["streamView"] = state ? state.streamView : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["tagsAll"] = state ? state.tagsAll : undefined;
            inputs["vpcConfig"] = state ? state.vpcConfig : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.computeCapacity === undefined) && !opts.urn) {
                throw new Error("Missing required property 'computeCapacity'");
            }
            if ((!args || args.instanceType === undefined) && !opts.urn) {
                throw new Error("Missing required property 'instanceType'");
            }
            inputs["computeCapacity"] = args ? args.computeCapacity : undefined;
            inputs["description"] = args ? args.description : undefined;
            inputs["disconnectTimeoutInSeconds"] = args ? args.disconnectTimeoutInSeconds : undefined;
            inputs["displayName"] = args ? args.displayName : undefined;
            inputs["domainJoinInfo"] = args ? args.domainJoinInfo : undefined;
            inputs["enableDefaultInternetAccess"] = args ? args.enableDefaultInternetAccess : undefined;
            inputs["fleetType"] = args ? args.fleetType : undefined;
            inputs["iamRoleArn"] = args ? args.iamRoleArn : undefined;
            inputs["idleDisconnectTimeoutInSeconds"] = args ? args.idleDisconnectTimeoutInSeconds : undefined;
            inputs["imageArn"] = args ? args.imageArn : undefined;
            inputs["imageName"] = args ? args.imageName : undefined;
            inputs["instanceType"] = args ? args.instanceType : undefined;
            inputs["maxUserDurationInSeconds"] = args ? args.maxUserDurationInSeconds : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["streamView"] = args ? args.streamView : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["vpcConfig"] = args ? args.vpcConfig : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["createdTime"] = undefined /*out*/;
            inputs["state"] = undefined /*out*/;
            inputs["tagsAll"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(Fleet.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Fleet resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Fleet(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Fleet.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Fleet.__pulumiType;
    }
}
exports.Fleet = Fleet;
/** @internal */
Fleet.__pulumiType = 'aws:appstream/fleet:Fleet';
