// @pulumi/aws/storagegateway/fileSystemAssociation.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileSystemAssociation = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Associate an Amazon FSx file system with the FSx File Gateway. After the association process is complete, the file shares on the Amazon FSx file system are available for access through the gateway. This operation only supports the FSx File Gateway type.
 *
 * [FSx File Gateway requirements](https://docs.aws.amazon.com/filegateway/latest/filefsxw/Requirements.html).
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.storagegateway.FileSystemAssociation("example", {
 *     gatewayArn: aws_storagegateway_gateway.example.arn,
 *     locationArn: aws_fsx_windows_file_system.example.arn,
 *     username: "Admin",
 *     password: "avoid-plaintext-passwords",
 *     auditDestinationArn: aws_s3_bucket.example.arn,
 * });
 * ```
 * ## Required Services Example
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const awsServiceStoragegatewayAmiFILES3Latest = aws.ssm.getParameter({
 *     name: "/aws/service/storagegateway/ami/FILE_S3/latest",
 * });
 * const testInstance = new aws.ec2.Instance("testInstance", {
 *     ami: awsServiceStoragegatewayAmiFILES3Latest.then(awsServiceStoragegatewayAmiFILES3Latest => awsServiceStoragegatewayAmiFILES3Latest.value),
 *     associatePublicIpAddress: true,
 *     instanceType: data.aws_ec2_instance_type_offering.available.instance_type,
 *     vpcSecurityGroupIds: [aws_security_group.test.id],
 *     subnetId: aws_subnet.test[0].id,
 * }, {
 *     dependsOn: [
 *         aws_route.test,
 *         aws_vpc_dhcp_options_association.test,
 *     ],
 * });
 * const testGateway = new aws.storagegateway.Gateway("testGateway", {
 *     gatewayIpAddress: testInstance.publicIp,
 *     gatewayName: "test-sgw",
 *     gatewayTimezone: "GMT",
 *     gatewayType: "FILE_FSX_SMB",
 *     smbActiveDirectorySettings: {
 *         domainName: aws_directory_service_directory.test.name,
 *         password: aws_directory_service_directory.test.password,
 *         username: "Admin",
 *     },
 * });
 * const testWindowsFileSystem = new aws.fsx.WindowsFileSystem("testWindowsFileSystem", {
 *     activeDirectoryId: aws_directory_service_directory.test.id,
 *     securityGroupIds: [aws_security_group.test.id],
 *     skipFinalBackup: true,
 *     storageCapacity: 32,
 *     subnetIds: [aws_subnet.test[0].id],
 *     throughputCapacity: 8,
 * });
 * const fsx = new aws.storagegateway.FileSystemAssociation("fsx", {
 *     gatewayArn: testGateway.arn,
 *     locationArn: testWindowsFileSystem.arn,
 *     username: "Admin",
 *     password: aws_directory_service_directory.test.password,
 *     cacheAttributes: {
 *         cacheStaleTimeoutInSeconds: 400,
 *     },
 *     auditDestinationArn: aws_cloudwatch_log_group.test.arn,
 * });
 * ```
 *
 * ## Import
 *
 * `aws_storagegateway_file_system_association` can be imported by using the FSx file system association Amazon Resource Name (ARN), e.g.,
 *
 * ```sh
 *  $ pulumi import aws:storagegateway/fileSystemAssociation:FileSystemAssociation example arn:aws:storagegateway:us-east-1:123456789012:fs-association/fsa-0DA347732FDB40125
 * ```
 */
class FileSystemAssociation extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["auditDestinationArn"] = state ? state.auditDestinationArn : undefined;
            inputs["cacheAttributes"] = state ? state.cacheAttributes : undefined;
            inputs["gatewayArn"] = state ? state.gatewayArn : undefined;
            inputs["locationArn"] = state ? state.locationArn : undefined;
            inputs["password"] = state ? state.password : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["tagsAll"] = state ? state.tagsAll : undefined;
            inputs["username"] = state ? state.username : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.gatewayArn === undefined) && !opts.urn) {
                throw new Error("Missing required property 'gatewayArn'");
            }
            if ((!args || args.locationArn === undefined) && !opts.urn) {
                throw new Error("Missing required property 'locationArn'");
            }
            if ((!args || args.password === undefined) && !opts.urn) {
                throw new Error("Missing required property 'password'");
            }
            if ((!args || args.username === undefined) && !opts.urn) {
                throw new Error("Missing required property 'username'");
            }
            inputs["auditDestinationArn"] = args ? args.auditDestinationArn : undefined;
            inputs["cacheAttributes"] = args ? args.cacheAttributes : undefined;
            inputs["gatewayArn"] = args ? args.gatewayArn : undefined;
            inputs["locationArn"] = args ? args.locationArn : undefined;
            inputs["password"] = args ? args.password : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["username"] = args ? args.username : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["tagsAll"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(FileSystemAssociation.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing FileSystemAssociation resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new FileSystemAssociation(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of FileSystemAssociation.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === FileSystemAssociation.__pulumiType;
    }
}
exports.FileSystemAssociation = FileSystemAssociation;
/** @internal */
FileSystemAssociation.__pulumiType = 'aws:storagegateway/fileSystemAssociation:FileSystemAssociation';
