// @pulumi/aws/servicecatalog/provisioningArtifact.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProvisioningArtifact = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Manages a Service Catalog Provisioning Artifact for a specified product.
 *
 * > A "provisioning artifact" is also referred to as a "version."
 *
 * > **NOTE:** You cannot create a provisioning artifact for a product that was shared with you.
 *
 * > **NOTE:** The user or role that use this resource must have the `cloudformation:GetTemplate` IAM policy permission. This policy permission is required when using the `templatePhysicalId` argument.
 *
 * ## Example Usage
 * ### Basic Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.servicecatalog.ProvisioningArtifact("example", {
 *     productId: aws_servicecatalog_product.example.id,
 *     type: "CLOUD_FORMATION_TEMPLATE",
 *     templateUrl: `https://${aws_s3_bucket.example.bucket_regional_domain_name}/${aws_s3_bucket_object.example.key}`,
 * });
 * ```
 *
 * ## Import
 *
 * `aws_servicecatalog_provisioning_artifact` can be imported using the provisioning artifact ID and product ID separated by a colon, e.g.,
 *
 * ```sh
 *  $ pulumi import aws:servicecatalog/provisioningArtifact:ProvisioningArtifact example pa-ij2b6lusy6dec:prod-el3an0rma3
 * ```
 */
class ProvisioningArtifact extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["acceptLanguage"] = state ? state.acceptLanguage : undefined;
            inputs["active"] = state ? state.active : undefined;
            inputs["createdTime"] = state ? state.createdTime : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["disableTemplateValidation"] = state ? state.disableTemplateValidation : undefined;
            inputs["guidance"] = state ? state.guidance : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["productId"] = state ? state.productId : undefined;
            inputs["templatePhysicalId"] = state ? state.templatePhysicalId : undefined;
            inputs["templateUrl"] = state ? state.templateUrl : undefined;
            inputs["type"] = state ? state.type : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.productId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'productId'");
            }
            inputs["acceptLanguage"] = args ? args.acceptLanguage : undefined;
            inputs["active"] = args ? args.active : undefined;
            inputs["description"] = args ? args.description : undefined;
            inputs["disableTemplateValidation"] = args ? args.disableTemplateValidation : undefined;
            inputs["guidance"] = args ? args.guidance : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["productId"] = args ? args.productId : undefined;
            inputs["templatePhysicalId"] = args ? args.templatePhysicalId : undefined;
            inputs["templateUrl"] = args ? args.templateUrl : undefined;
            inputs["type"] = args ? args.type : undefined;
            inputs["createdTime"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(ProvisioningArtifact.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing ProvisioningArtifact resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ProvisioningArtifact(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ProvisioningArtifact.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ProvisioningArtifact.__pulumiType;
    }
}
exports.ProvisioningArtifact = ProvisioningArtifact;
/** @internal */
ProvisioningArtifact.__pulumiType = 'aws:servicecatalog/provisioningArtifact:ProvisioningArtifact';
