// @pulumi/aws/s3control/objectLambdaAccessPointPolicy.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectLambdaAccessPointPolicy = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a resource to manage an S3 Object Lambda Access Point resource policy.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const exampleBucket = new aws.s3.Bucket("exampleBucket", {});
 * const exampleAccessPoint = new aws.s3.AccessPoint("exampleAccessPoint", {bucket: exampleBucket.id});
 * const exampleObjectLambdaAccessPoint = new aws.s3control.ObjectLambdaAccessPoint("exampleObjectLambdaAccessPoint", {configuration: {
 *     supportingAccessPoint: exampleAccessPoint.arn,
 *     transformationConfigurations: [{
 *         actions: ["GetObject"],
 *         contentTransformation: {
 *             awsLambda: {
 *                 functionArn: aws_lambda_function.example.arn,
 *             },
 *         },
 *     }],
 * }});
 * const exampleObjectLambdaAccessPointPolicy = new aws.s3control.ObjectLambdaAccessPointPolicy("exampleObjectLambdaAccessPointPolicy", {policy: exampleObjectLambdaAccessPoint.arn.apply(arn => JSON.stringify({
 *     Version: "2008-10-17",
 *     Statement: [{
 *         Effect: "Allow",
 *         Action: "s3-object-lambda:GetObject",
 *         Principal: {
 *             AWS: data.aws_caller_identity.current.account_id,
 *         },
 *         Resource: arn,
 *     }],
 * }))});
 * ```
 *
 * ## Import
 *
 * Object Lambda Access Point policies can be imported using the `account_id` and `name`, separated by a colon (`:`), e.g.
 *
 * ```sh
 *  $ pulumi import aws:s3control/objectLambdaAccessPointPolicy:ObjectLambdaAccessPointPolicy example 123456789012:example
 * ```
 */
class ObjectLambdaAccessPointPolicy extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["accountId"] = state ? state.accountId : undefined;
            inputs["hasPublicAccessPolicy"] = state ? state.hasPublicAccessPolicy : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["policy"] = state ? state.policy : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.policy === undefined) && !opts.urn) {
                throw new Error("Missing required property 'policy'");
            }
            inputs["accountId"] = args ? args.accountId : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["policy"] = args ? args.policy : undefined;
            inputs["hasPublicAccessPolicy"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(ObjectLambdaAccessPointPolicy.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing ObjectLambdaAccessPointPolicy resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ObjectLambdaAccessPointPolicy(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ObjectLambdaAccessPointPolicy.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ObjectLambdaAccessPointPolicy.__pulumiType;
    }
}
exports.ObjectLambdaAccessPointPolicy = ObjectLambdaAccessPointPolicy;
/** @internal */
ObjectLambdaAccessPointPolicy.__pulumiType = 'aws:s3control/objectLambdaAccessPointPolicy:ObjectLambdaAccessPointPolicy';
