// @pulumi/aws/securityhub/standardsControl.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.StandardsControl = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
class StandardsControl extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["controlId"] = state ? state.controlId : undefined;
            inputs["controlStatus"] = state ? state.controlStatus : undefined;
            inputs["controlStatusUpdatedAt"] = state ? state.controlStatusUpdatedAt : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["disabledReason"] = state ? state.disabledReason : undefined;
            inputs["relatedRequirements"] = state ? state.relatedRequirements : undefined;
            inputs["remediationUrl"] = state ? state.remediationUrl : undefined;
            inputs["severityRating"] = state ? state.severityRating : undefined;
            inputs["standardsControlArn"] = state ? state.standardsControlArn : undefined;
            inputs["title"] = state ? state.title : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.controlStatus === undefined) && !opts.urn) {
                throw new Error("Missing required property 'controlStatus'");
            }
            if ((!args || args.standardsControlArn === undefined) && !opts.urn) {
                throw new Error("Missing required property 'standardsControlArn'");
            }
            inputs["controlStatus"] = args ? args.controlStatus : undefined;
            inputs["disabledReason"] = args ? args.disabledReason : undefined;
            inputs["standardsControlArn"] = args ? args.standardsControlArn : undefined;
            inputs["controlId"] = undefined /*out*/;
            inputs["controlStatusUpdatedAt"] = undefined /*out*/;
            inputs["description"] = undefined /*out*/;
            inputs["relatedRequirements"] = undefined /*out*/;
            inputs["remediationUrl"] = undefined /*out*/;
            inputs["severityRating"] = undefined /*out*/;
            inputs["title"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(StandardsControl.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing StandardsControl resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new StandardsControl(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of StandardsControl.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === StandardsControl.__pulumiType;
    }
}
exports.StandardsControl = StandardsControl;
/** @internal */
StandardsControl.__pulumiType = 'aws:securityhub/standardsControl:StandardsControl';
