// @pulumi/aws/rds/proxyEndpoint.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProxyEndpoint = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides an RDS DB proxy endpoint resource. For additional information, see the [RDS User Guide](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/rds-proxy.html#rds-proxy-endpoints).
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.rds.ProxyEndpoint("example", {
 *     dbProxyName: aws_db_proxy.test.name,
 *     dbProxyEndpointName: "example",
 *     vpcSubnetIds: aws_subnet.test.map(__item => __item.id),
 *     targetRole: "READ_ONLY",
 * });
 * ```
 *
 * ## Import
 *
 * DB proxy endpoints can be imported using the `DB-PROXY-NAME/DB-PROXY-ENDPOINT-NAME`, e.g.,
 *
 * ```sh
 *  $ pulumi import aws:rds/proxyEndpoint:ProxyEndpoint example example/example
 * ```
 */
class ProxyEndpoint extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["dbProxyEndpointName"] = state ? state.dbProxyEndpointName : undefined;
            inputs["dbProxyName"] = state ? state.dbProxyName : undefined;
            inputs["endpoint"] = state ? state.endpoint : undefined;
            inputs["isDefault"] = state ? state.isDefault : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["tagsAll"] = state ? state.tagsAll : undefined;
            inputs["targetRole"] = state ? state.targetRole : undefined;
            inputs["vpcId"] = state ? state.vpcId : undefined;
            inputs["vpcSecurityGroupIds"] = state ? state.vpcSecurityGroupIds : undefined;
            inputs["vpcSubnetIds"] = state ? state.vpcSubnetIds : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.dbProxyEndpointName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'dbProxyEndpointName'");
            }
            if ((!args || args.dbProxyName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'dbProxyName'");
            }
            if ((!args || args.vpcSubnetIds === undefined) && !opts.urn) {
                throw new Error("Missing required property 'vpcSubnetIds'");
            }
            inputs["dbProxyEndpointName"] = args ? args.dbProxyEndpointName : undefined;
            inputs["dbProxyName"] = args ? args.dbProxyName : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["targetRole"] = args ? args.targetRole : undefined;
            inputs["vpcSecurityGroupIds"] = args ? args.vpcSecurityGroupIds : undefined;
            inputs["vpcSubnetIds"] = args ? args.vpcSubnetIds : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["endpoint"] = undefined /*out*/;
            inputs["isDefault"] = undefined /*out*/;
            inputs["tagsAll"] = undefined /*out*/;
            inputs["vpcId"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(ProxyEndpoint.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing ProxyEndpoint resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ProxyEndpoint(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ProxyEndpoint.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ProxyEndpoint.__pulumiType;
    }
}
exports.ProxyEndpoint = ProxyEndpoint;
/** @internal */
ProxyEndpoint.__pulumiType = 'aws:rds/proxyEndpoint:ProxyEndpoint';
