// @pulumi/aws/route53/resolverFirewallRule.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResolverFirewallRule = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a Route 53 Resolver DNS Firewall rule resource.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const exampleResolverFirewallDomainList = new aws.route53.ResolverFirewallDomainList("exampleResolverFirewallDomainList", {
 *     domains: ["example.com"],
 *     tags: {},
 * });
 * const exampleResolverFirewallRuleGroup = new aws.route53.ResolverFirewallRuleGroup("exampleResolverFirewallRuleGroup", {tags: {}});
 * const exampleResolverFirewallRule = new aws.route53.ResolverFirewallRule("exampleResolverFirewallRule", {
 *     action: "BLOCK",
 *     blockOverrideDnsType: "CNAME",
 *     blockOverrideDomain: "example.com",
 *     blockOverrideTtl: 1,
 *     blockResponse: "OVERRIDE",
 *     firewallDomainListId: exampleResolverFirewallDomainList.id,
 *     firewallRuleGroupId: exampleResolverFirewallRuleGroup.id,
 *     priority: 100,
 * });
 * ```
 *
 * ## Import
 *
 *  Route 53 Resolver DNS Firewall rules can be imported using the Route 53 Resolver DNS Firewall rule group ID and domain list ID separated by ':', e.g.,
 *
 * ```sh
 *  $ pulumi import aws:route53/resolverFirewallRule:ResolverFirewallRule example rslvr-frg-0123456789abcdef:rslvr-fdl-0123456789abcdef
 * ```
 */
class ResolverFirewallRule extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["action"] = state ? state.action : undefined;
            inputs["blockOverrideDnsType"] = state ? state.blockOverrideDnsType : undefined;
            inputs["blockOverrideDomain"] = state ? state.blockOverrideDomain : undefined;
            inputs["blockOverrideTtl"] = state ? state.blockOverrideTtl : undefined;
            inputs["blockResponse"] = state ? state.blockResponse : undefined;
            inputs["firewallDomainListId"] = state ? state.firewallDomainListId : undefined;
            inputs["firewallRuleGroupId"] = state ? state.firewallRuleGroupId : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["priority"] = state ? state.priority : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.action === undefined) && !opts.urn) {
                throw new Error("Missing required property 'action'");
            }
            if ((!args || args.firewallDomainListId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'firewallDomainListId'");
            }
            if ((!args || args.firewallRuleGroupId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'firewallRuleGroupId'");
            }
            if ((!args || args.priority === undefined) && !opts.urn) {
                throw new Error("Missing required property 'priority'");
            }
            inputs["action"] = args ? args.action : undefined;
            inputs["blockOverrideDnsType"] = args ? args.blockOverrideDnsType : undefined;
            inputs["blockOverrideDomain"] = args ? args.blockOverrideDomain : undefined;
            inputs["blockOverrideTtl"] = args ? args.blockOverrideTtl : undefined;
            inputs["blockResponse"] = args ? args.blockResponse : undefined;
            inputs["firewallDomainListId"] = args ? args.firewallDomainListId : undefined;
            inputs["firewallRuleGroupId"] = args ? args.firewallRuleGroupId : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["priority"] = args ? args.priority : undefined;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(ResolverFirewallRule.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing ResolverFirewallRule resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ResolverFirewallRule(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ResolverFirewallRule.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ResolverFirewallRule.__pulumiType;
    }
}
exports.ResolverFirewallRule = ResolverFirewallRule;
/** @internal */
ResolverFirewallRule.__pulumiType = 'aws:route53/resolverFirewallRule:ResolverFirewallRule';
