// @pulumi/aws/cloudformation/cloudFormationType.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudFormationType = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * ## Import
 *
 * `aws_cloudformation_type` can be imported with their type version Amazon Resource Name (ARN), e.g.,
 *
 * ```sh
 *  $ pulumi import aws:cloudformation/cloudFormationType:CloudFormationType example arn:aws:cloudformation:us-east-1:123456789012:type/resource/ExampleCompany-ExampleService-ExampleType/1
 * ```
 */
class CloudFormationType extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["defaultVersionId"] = state ? state.defaultVersionId : undefined;
            inputs["deprecatedStatus"] = state ? state.deprecatedStatus : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["documentationUrl"] = state ? state.documentationUrl : undefined;
            inputs["executionRoleArn"] = state ? state.executionRoleArn : undefined;
            inputs["isDefaultVersion"] = state ? state.isDefaultVersion : undefined;
            inputs["loggingConfig"] = state ? state.loggingConfig : undefined;
            inputs["provisioningType"] = state ? state.provisioningType : undefined;
            inputs["schema"] = state ? state.schema : undefined;
            inputs["schemaHandlerPackage"] = state ? state.schemaHandlerPackage : undefined;
            inputs["sourceUrl"] = state ? state.sourceUrl : undefined;
            inputs["type"] = state ? state.type : undefined;
            inputs["typeArn"] = state ? state.typeArn : undefined;
            inputs["typeName"] = state ? state.typeName : undefined;
            inputs["versionId"] = state ? state.versionId : undefined;
            inputs["visibility"] = state ? state.visibility : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.schemaHandlerPackage === undefined) && !opts.urn) {
                throw new Error("Missing required property 'schemaHandlerPackage'");
            }
            if ((!args || args.typeName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'typeName'");
            }
            inputs["executionRoleArn"] = args ? args.executionRoleArn : undefined;
            inputs["loggingConfig"] = args ? args.loggingConfig : undefined;
            inputs["schemaHandlerPackage"] = args ? args.schemaHandlerPackage : undefined;
            inputs["type"] = args ? args.type : undefined;
            inputs["typeName"] = args ? args.typeName : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["defaultVersionId"] = undefined /*out*/;
            inputs["deprecatedStatus"] = undefined /*out*/;
            inputs["description"] = undefined /*out*/;
            inputs["documentationUrl"] = undefined /*out*/;
            inputs["isDefaultVersion"] = undefined /*out*/;
            inputs["provisioningType"] = undefined /*out*/;
            inputs["schema"] = undefined /*out*/;
            inputs["sourceUrl"] = undefined /*out*/;
            inputs["typeArn"] = undefined /*out*/;
            inputs["versionId"] = undefined /*out*/;
            inputs["visibility"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(CloudFormationType.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing CloudFormationType resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new CloudFormationType(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of CloudFormationType.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === CloudFormationType.__pulumiType;
    }
}
exports.CloudFormationType = CloudFormationType;
/** @internal */
CloudFormationType.__pulumiType = 'aws:cloudformation/cloudFormationType:CloudFormationType';
