// @pulumi/aws/macie2/classificationJob.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClassificationJob = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a resource to manage an [AWS Macie Classification Job](https://docs.aws.amazon.com/macie/latest/APIReference/jobs.html).
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const testAccount = new aws.macie2.Account("testAccount", {});
 * const testClassificationJob = new aws.macie2.ClassificationJob("testClassificationJob", {
 *     jobType: "ONE_TIME",
 *     s3JobDefinition: {
 *         bucketDefinitions: [{
 *             accountId: "ACCOUNT ID",
 *             buckets: ["S3 BUCKET NAME"],
 *         }],
 *     },
 * }, {
 *     dependsOn: [testAccount],
 * });
 * ```
 *
 * ## Import
 *
 * `aws_macie2_classification_job` can be imported using the id, e.g.,
 *
 * ```sh
 *  $ pulumi import aws:macie2/classificationJob:ClassificationJob example abcd1
 * ```
 */
class ClassificationJob extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["createdAt"] = state ? state.createdAt : undefined;
            inputs["customDataIdentifierIds"] = state ? state.customDataIdentifierIds : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["initialRun"] = state ? state.initialRun : undefined;
            inputs["jobArn"] = state ? state.jobArn : undefined;
            inputs["jobId"] = state ? state.jobId : undefined;
            inputs["jobStatus"] = state ? state.jobStatus : undefined;
            inputs["jobType"] = state ? state.jobType : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["namePrefix"] = state ? state.namePrefix : undefined;
            inputs["s3JobDefinition"] = state ? state.s3JobDefinition : undefined;
            inputs["samplingPercentage"] = state ? state.samplingPercentage : undefined;
            inputs["scheduleFrequency"] = state ? state.scheduleFrequency : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["tagsAll"] = state ? state.tagsAll : undefined;
            inputs["userPausedDetails"] = state ? state.userPausedDetails : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.jobType === undefined) && !opts.urn) {
                throw new Error("Missing required property 'jobType'");
            }
            if ((!args || args.s3JobDefinition === undefined) && !opts.urn) {
                throw new Error("Missing required property 's3JobDefinition'");
            }
            inputs["customDataIdentifierIds"] = args ? args.customDataIdentifierIds : undefined;
            inputs["description"] = args ? args.description : undefined;
            inputs["initialRun"] = args ? args.initialRun : undefined;
            inputs["jobStatus"] = args ? args.jobStatus : undefined;
            inputs["jobType"] = args ? args.jobType : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["namePrefix"] = args ? args.namePrefix : undefined;
            inputs["s3JobDefinition"] = args ? args.s3JobDefinition : undefined;
            inputs["samplingPercentage"] = args ? args.samplingPercentage : undefined;
            inputs["scheduleFrequency"] = args ? args.scheduleFrequency : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["createdAt"] = undefined /*out*/;
            inputs["jobArn"] = undefined /*out*/;
            inputs["jobId"] = undefined /*out*/;
            inputs["tagsAll"] = undefined /*out*/;
            inputs["userPausedDetails"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(ClassificationJob.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing ClassificationJob resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ClassificationJob(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ClassificationJob.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ClassificationJob.__pulumiType;
    }
}
exports.ClassificationJob = ClassificationJob;
/** @internal */
ClassificationJob.__pulumiType = 'aws:macie2/classificationJob:ClassificationJob';
