// @pulumi/aws/s3/objectCopy.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectCopy = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a resource for copying an S3 object.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const test = new aws.s3.ObjectCopy("test", {
 *     bucket: "destination_bucket",
 *     grants: [{
 *         permissions: ["READ"],
 *         type: "Group",
 *         uri: "http://acs.amazonaws.com/groups/global/AllUsers",
 *     }],
 *     key: "destination_key",
 *     source: "source_bucket/source_key",
 * });
 * ```
 */
class ObjectCopy extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["acl"] = state ? state.acl : undefined;
            inputs["bucket"] = state ? state.bucket : undefined;
            inputs["bucketKeyEnabled"] = state ? state.bucketKeyEnabled : undefined;
            inputs["cacheControl"] = state ? state.cacheControl : undefined;
            inputs["contentDisposition"] = state ? state.contentDisposition : undefined;
            inputs["contentEncoding"] = state ? state.contentEncoding : undefined;
            inputs["contentLanguage"] = state ? state.contentLanguage : undefined;
            inputs["contentType"] = state ? state.contentType : undefined;
            inputs["copyIfMatch"] = state ? state.copyIfMatch : undefined;
            inputs["copyIfModifiedSince"] = state ? state.copyIfModifiedSince : undefined;
            inputs["copyIfNoneMatch"] = state ? state.copyIfNoneMatch : undefined;
            inputs["copyIfUnmodifiedSince"] = state ? state.copyIfUnmodifiedSince : undefined;
            inputs["customerAlgorithm"] = state ? state.customerAlgorithm : undefined;
            inputs["customerKey"] = state ? state.customerKey : undefined;
            inputs["customerKeyMd5"] = state ? state.customerKeyMd5 : undefined;
            inputs["etag"] = state ? state.etag : undefined;
            inputs["expectedBucketOwner"] = state ? state.expectedBucketOwner : undefined;
            inputs["expectedSourceBucketOwner"] = state ? state.expectedSourceBucketOwner : undefined;
            inputs["expiration"] = state ? state.expiration : undefined;
            inputs["expires"] = state ? state.expires : undefined;
            inputs["forceDestroy"] = state ? state.forceDestroy : undefined;
            inputs["grants"] = state ? state.grants : undefined;
            inputs["key"] = state ? state.key : undefined;
            inputs["kmsEncryptionContext"] = state ? state.kmsEncryptionContext : undefined;
            inputs["kmsKeyId"] = state ? state.kmsKeyId : undefined;
            inputs["lastModified"] = state ? state.lastModified : undefined;
            inputs["metadata"] = state ? state.metadata : undefined;
            inputs["metadataDirective"] = state ? state.metadataDirective : undefined;
            inputs["objectLockLegalHoldStatus"] = state ? state.objectLockLegalHoldStatus : undefined;
            inputs["objectLockMode"] = state ? state.objectLockMode : undefined;
            inputs["objectLockRetainUntilDate"] = state ? state.objectLockRetainUntilDate : undefined;
            inputs["requestCharged"] = state ? state.requestCharged : undefined;
            inputs["requestPayer"] = state ? state.requestPayer : undefined;
            inputs["serverSideEncryption"] = state ? state.serverSideEncryption : undefined;
            inputs["source"] = state ? state.source : undefined;
            inputs["sourceCustomerAlgorithm"] = state ? state.sourceCustomerAlgorithm : undefined;
            inputs["sourceCustomerKey"] = state ? state.sourceCustomerKey : undefined;
            inputs["sourceCustomerKeyMd5"] = state ? state.sourceCustomerKeyMd5 : undefined;
            inputs["sourceVersionId"] = state ? state.sourceVersionId : undefined;
            inputs["storageClass"] = state ? state.storageClass : undefined;
            inputs["taggingDirective"] = state ? state.taggingDirective : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["tagsAll"] = state ? state.tagsAll : undefined;
            inputs["versionId"] = state ? state.versionId : undefined;
            inputs["websiteRedirect"] = state ? state.websiteRedirect : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.bucket === undefined) && !opts.urn) {
                throw new Error("Missing required property 'bucket'");
            }
            if ((!args || args.key === undefined) && !opts.urn) {
                throw new Error("Missing required property 'key'");
            }
            if ((!args || args.source === undefined) && !opts.urn) {
                throw new Error("Missing required property 'source'");
            }
            inputs["acl"] = args ? args.acl : undefined;
            inputs["bucket"] = args ? args.bucket : undefined;
            inputs["bucketKeyEnabled"] = args ? args.bucketKeyEnabled : undefined;
            inputs["cacheControl"] = args ? args.cacheControl : undefined;
            inputs["contentDisposition"] = args ? args.contentDisposition : undefined;
            inputs["contentEncoding"] = args ? args.contentEncoding : undefined;
            inputs["contentLanguage"] = args ? args.contentLanguage : undefined;
            inputs["contentType"] = args ? args.contentType : undefined;
            inputs["copyIfMatch"] = args ? args.copyIfMatch : undefined;
            inputs["copyIfModifiedSince"] = args ? args.copyIfModifiedSince : undefined;
            inputs["copyIfNoneMatch"] = args ? args.copyIfNoneMatch : undefined;
            inputs["copyIfUnmodifiedSince"] = args ? args.copyIfUnmodifiedSince : undefined;
            inputs["customerAlgorithm"] = args ? args.customerAlgorithm : undefined;
            inputs["customerKey"] = args ? args.customerKey : undefined;
            inputs["customerKeyMd5"] = args ? args.customerKeyMd5 : undefined;
            inputs["expectedBucketOwner"] = args ? args.expectedBucketOwner : undefined;
            inputs["expectedSourceBucketOwner"] = args ? args.expectedSourceBucketOwner : undefined;
            inputs["expires"] = args ? args.expires : undefined;
            inputs["forceDestroy"] = args ? args.forceDestroy : undefined;
            inputs["grants"] = args ? args.grants : undefined;
            inputs["key"] = args ? args.key : undefined;
            inputs["kmsEncryptionContext"] = args ? args.kmsEncryptionContext : undefined;
            inputs["kmsKeyId"] = args ? args.kmsKeyId : undefined;
            inputs["metadata"] = args ? args.metadata : undefined;
            inputs["metadataDirective"] = args ? args.metadataDirective : undefined;
            inputs["objectLockLegalHoldStatus"] = args ? args.objectLockLegalHoldStatus : undefined;
            inputs["objectLockMode"] = args ? args.objectLockMode : undefined;
            inputs["objectLockRetainUntilDate"] = args ? args.objectLockRetainUntilDate : undefined;
            inputs["requestPayer"] = args ? args.requestPayer : undefined;
            inputs["serverSideEncryption"] = args ? args.serverSideEncryption : undefined;
            inputs["source"] = args ? args.source : undefined;
            inputs["sourceCustomerAlgorithm"] = args ? args.sourceCustomerAlgorithm : undefined;
            inputs["sourceCustomerKey"] = args ? args.sourceCustomerKey : undefined;
            inputs["sourceCustomerKeyMd5"] = args ? args.sourceCustomerKeyMd5 : undefined;
            inputs["storageClass"] = args ? args.storageClass : undefined;
            inputs["taggingDirective"] = args ? args.taggingDirective : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["websiteRedirect"] = args ? args.websiteRedirect : undefined;
            inputs["etag"] = undefined /*out*/;
            inputs["expiration"] = undefined /*out*/;
            inputs["lastModified"] = undefined /*out*/;
            inputs["requestCharged"] = undefined /*out*/;
            inputs["sourceVersionId"] = undefined /*out*/;
            inputs["tagsAll"] = undefined /*out*/;
            inputs["versionId"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(ObjectCopy.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing ObjectCopy resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ObjectCopy(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ObjectCopy.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ObjectCopy.__pulumiType;
    }
}
exports.ObjectCopy = ObjectCopy;
/** @internal */
ObjectCopy.__pulumiType = 'aws:s3/objectCopy:ObjectCopy';
