// @pulumi/aws/sagemaker/humanTaskUI.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.HumanTaskUI = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a Sagemaker Human Task UI resource.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 * import * from "fs";
 *
 * const example = new aws.sagemaker.HumanTaskUI("example", {
 *     humanTaskUiName: "example",
 *     uiTemplate: {
 *         content: fs.readFileSync("sagemaker-human-task-ui-template.html"),
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * Sagemaker Human Task UIs can be imported using the `human_task_ui_name`, e.g.,
 *
 * ```sh
 *  $ pulumi import aws:sagemaker/humanTaskUI:HumanTaskUI example example
 * ```
 */
class HumanTaskUI extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["humanTaskUiName"] = state ? state.humanTaskUiName : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["tagsAll"] = state ? state.tagsAll : undefined;
            inputs["uiTemplate"] = state ? state.uiTemplate : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.humanTaskUiName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'humanTaskUiName'");
            }
            if ((!args || args.uiTemplate === undefined) && !opts.urn) {
                throw new Error("Missing required property 'uiTemplate'");
            }
            inputs["humanTaskUiName"] = args ? args.humanTaskUiName : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["uiTemplate"] = args ? args.uiTemplate : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["tagsAll"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(HumanTaskUI.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing HumanTaskUI resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new HumanTaskUI(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of HumanTaskUI.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === HumanTaskUI.__pulumiType;
    }
}
exports.HumanTaskUI = HumanTaskUI;
/** @internal */
HumanTaskUI.__pulumiType = 'aws:sagemaker/humanTaskUI:HumanTaskUI';
