// @pulumi/aws/sagemaker/modelPackageGroup.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModelPackageGroup = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a Sagemaker Model Package Group resource.
 *
 * ## Example Usage
 * ### Basic usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.sagemaker.ModelPackageGroup("example", {
 *     modelPackageGroupName: "example",
 * });
 * ```
 *
 * ## Import
 *
 * Sagemaker Code Model Package Groups can be imported using the `name`, e.g.,
 *
 * ```sh
 *  $ pulumi import aws:sagemaker/modelPackageGroup:ModelPackageGroup test_model_package_group my-code-repo
 * ```
 */
class ModelPackageGroup extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["modelPackageGroupDescription"] = state ? state.modelPackageGroupDescription : undefined;
            inputs["modelPackageGroupName"] = state ? state.modelPackageGroupName : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["tagsAll"] = state ? state.tagsAll : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.modelPackageGroupName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'modelPackageGroupName'");
            }
            inputs["modelPackageGroupDescription"] = args ? args.modelPackageGroupDescription : undefined;
            inputs["modelPackageGroupName"] = args ? args.modelPackageGroupName : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["tagsAll"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(ModelPackageGroup.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing ModelPackageGroup resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ModelPackageGroup(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ModelPackageGroup.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ModelPackageGroup.__pulumiType;
    }
}
exports.ModelPackageGroup = ModelPackageGroup;
/** @internal */
ModelPackageGroup.__pulumiType = 'aws:sagemaker/modelPackageGroup:ModelPackageGroup';
