// @pulumi/aws/sagemaker/userProfile.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserProfile = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a Sagemaker User Profile resource.
 *
 * ## Example Usage
 * ### Basic usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.sagemaker.UserProfile("example", {
 *     domainId: aws_sagemaker_domain.test.id,
 *     userProfileName: "example",
 * });
 * ```
 *
 * ## Import
 *
 * Sagemaker Code User Profiles can be imported using the `arn`, e.g.,
 *
 * ```sh
 *  $ pulumi import aws:sagemaker/userProfile:UserProfile test_user_profile arn:aws:sagemaker:us-west-2:123456789012:user-profile/domain-id/profile-name
 * ```
 */
class UserProfile extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["domainId"] = state ? state.domainId : undefined;
            inputs["homeEfsFileSystemUid"] = state ? state.homeEfsFileSystemUid : undefined;
            inputs["singleSignOnUserIdentifier"] = state ? state.singleSignOnUserIdentifier : undefined;
            inputs["singleSignOnUserValue"] = state ? state.singleSignOnUserValue : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["tagsAll"] = state ? state.tagsAll : undefined;
            inputs["userProfileName"] = state ? state.userProfileName : undefined;
            inputs["userSettings"] = state ? state.userSettings : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.domainId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'domainId'");
            }
            if ((!args || args.userProfileName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'userProfileName'");
            }
            inputs["domainId"] = args ? args.domainId : undefined;
            inputs["singleSignOnUserIdentifier"] = args ? args.singleSignOnUserIdentifier : undefined;
            inputs["singleSignOnUserValue"] = args ? args.singleSignOnUserValue : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["userProfileName"] = args ? args.userProfileName : undefined;
            inputs["userSettings"] = args ? args.userSettings : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["homeEfsFileSystemUid"] = undefined /*out*/;
            inputs["tagsAll"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(UserProfile.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing UserProfile resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new UserProfile(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of UserProfile.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === UserProfile.__pulumiType;
    }
}
exports.UserProfile = UserProfile;
/** @internal */
UserProfile.__pulumiType = 'aws:sagemaker/userProfile:UserProfile';
