// @pulumi/aws/kms/replicaExternalKey.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReplicaExternalKey = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Manages a KMS multi-Region replica key that uses external key material.
 * See the [AWS KMS Developer Guide](https://docs.aws.amazon.com/kms/latest/developerguide/multi-region-keys-import.html) for more information on importing key material into multi-Region keys.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const primary = new aws.Provider("primary", {region: "us-east-1"});
 * const primaryExternalKey = new aws.kms.ExternalKey("primaryExternalKey", {
 *     description: "Multi-Region primary key",
 *     deletionWindowInDays: 30,
 *     multiRegion: true,
 *     enabled: true,
 *     keyMaterialBase64: "...",
 * }, {
 *     provider: aws.primary,
 * });
 * const replica = new aws.kms.ReplicaExternalKey("replica", {
 *     description: "Multi-Region replica key",
 *     deletionWindowInDays: 7,
 *     primaryKeyArn: aws_kms_external.primary.arn,
 *     keyMaterialBase64: "...",
 * });
 * // Must be the same key material as the primary's.
 * ```
 *
 * ## Import
 *
 * KMS multi-Region replica keys can be imported using the `id`, e.g.,
 *
 * ```sh
 *  $ pulumi import aws:kms/replicaExternalKey:ReplicaExternalKey example 1234abcd-12ab-34cd-56ef-1234567890ab
 * ```
 */
class ReplicaExternalKey extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["bypassPolicyLockoutSafetyCheck"] = state ? state.bypassPolicyLockoutSafetyCheck : undefined;
            inputs["deletionWindowInDays"] = state ? state.deletionWindowInDays : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["enabled"] = state ? state.enabled : undefined;
            inputs["expirationModel"] = state ? state.expirationModel : undefined;
            inputs["keyId"] = state ? state.keyId : undefined;
            inputs["keyMaterialBase64"] = state ? state.keyMaterialBase64 : undefined;
            inputs["keyState"] = state ? state.keyState : undefined;
            inputs["keyUsage"] = state ? state.keyUsage : undefined;
            inputs["policy"] = state ? state.policy : undefined;
            inputs["primaryKeyArn"] = state ? state.primaryKeyArn : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["tagsAll"] = state ? state.tagsAll : undefined;
            inputs["validTo"] = state ? state.validTo : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.primaryKeyArn === undefined) && !opts.urn) {
                throw new Error("Missing required property 'primaryKeyArn'");
            }
            inputs["bypassPolicyLockoutSafetyCheck"] = args ? args.bypassPolicyLockoutSafetyCheck : undefined;
            inputs["deletionWindowInDays"] = args ? args.deletionWindowInDays : undefined;
            inputs["description"] = args ? args.description : undefined;
            inputs["enabled"] = args ? args.enabled : undefined;
            inputs["keyMaterialBase64"] = args ? args.keyMaterialBase64 : undefined;
            inputs["policy"] = args ? args.policy : undefined;
            inputs["primaryKeyArn"] = args ? args.primaryKeyArn : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["validTo"] = args ? args.validTo : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["expirationModel"] = undefined /*out*/;
            inputs["keyId"] = undefined /*out*/;
            inputs["keyState"] = undefined /*out*/;
            inputs["keyUsage"] = undefined /*out*/;
            inputs["tagsAll"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(ReplicaExternalKey.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing ReplicaExternalKey resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ReplicaExternalKey(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ReplicaExternalKey.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ReplicaExternalKey.__pulumiType;
    }
}
exports.ReplicaExternalKey = ReplicaExternalKey;
/** @internal */
ReplicaExternalKey.__pulumiType = 'aws:kms/replicaExternalKey:ReplicaExternalKey';
