// @pulumi/aws/amplify/branch.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Branch = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides an Amplify Branch resource.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.amplify.App("example", {});
 * const master = new aws.amplify.Branch("master", {
 *     appId: example.id,
 *     branchName: "master",
 *     framework: "React",
 *     stage: "PRODUCTION",
 *     environmentVariables: {
 *         REACT_APP_API_SERVER: "https://api.example.com",
 *     },
 * });
 * ```
 * ### Basic Authentication
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.amplify.App("example", {});
 * const master = new aws.amplify.Branch("master", {
 *     appId: example.id,
 *     branchName: "master",
 *     basicAuthConfig: [{
 *         enableBasicAuth: true,
 *         username: "username",
 *         password: "password",
 *     }],
 * });
 * ```
 * ### Notifications
 *
 * Amplify Console uses EventBridge (formerly known as CloudWatch Events) and SNS for email notifications.  To implement the same functionality, you need to set `enableNotification` in a `aws.amplify.Branch` resource, as well as creating an EventBridge Rule, an SNS topic, and SNS subscriptions.
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.amplify.App("example", {});
 * const master = new aws.amplify.Branch("master", {
 *     appId: example.id,
 *     branchName: "master",
 *     enableNotification: true,
 * });
 * // EventBridge Rule for Amplify notifications
 * const amplifyAppMasterEventRule = new aws.cloudwatch.EventRule("amplifyAppMasterEventRule", {
 *     description: pulumi.interpolate`AWS Amplify build notifications for :  App: ${aws_amplify_app.app.id} Branch: ${master.branchName}`,
 *     eventPattern: pulumi.all([example.id, master.branchName]).apply(([id, branchName]) => JSON.stringify({
 *         detail: {
 *             appId: [id],
 *             branchName: [branchName],
 *             jobStatus: [
 *                 "SUCCEED",
 *                 "FAILED",
 *                 "STARTED",
 *             ],
 *         },
 *         "detail-type": ["Amplify Deployment Status Change"],
 *         source: ["aws.amplify"],
 *     })),
 * });
 * const amplifyAppMasterTopic = new aws.sns.Topic("amplifyAppMasterTopic", {});
 * const amplifyAppMasterEventTarget = new aws.cloudwatch.EventTarget("amplifyAppMasterEventTarget", {
 *     rule: amplifyAppMasterEventRule.name,
 *     arn: amplifyAppMasterTopic.arn,
 *     inputTransformer: {
 *         inputPaths: {
 *             jobId: `$.detail.jobId`,
 *             appId: `$.detail.appId`,
 *             region: `$.region`,
 *             branch: `$.detail.branchName`,
 *             status: `$.detail.jobStatus`,
 *         },
 *         inputTemplate: "\"Build notification from the AWS Amplify Console for app: https://<branch>.<appId>.amplifyapp.com/. Your build status is <status>. Go to https://console.aws.amazon.com/amplify/home?region=<region>#<appId>/<branch>/<jobId> to view details on your build. \"",
 *     },
 * });
 * // SNS Topic for Amplify notifications
 * const amplifyAppMasterPolicyDocument = pulumi.all([master.arn, amplifyAppMasterTopic.arn]).apply(([masterArn, amplifyAppMasterTopicArn]) => aws.iam.getPolicyDocument({
 *     statements: [{
 *         sid: `Allow_Publish_Events ${masterArn}`,
 *         effect: "Allow",
 *         actions: ["SNS:Publish"],
 *         principals: [{
 *             type: "Service",
 *             identifiers: ["events.amazonaws.com"],
 *         }],
 *         resources: [amplifyAppMasterTopicArn],
 *     }],
 * }));
 * const amplifyAppMasterTopicPolicy = new aws.sns.TopicPolicy("amplifyAppMasterTopicPolicy", {
 *     arn: amplifyAppMasterTopic.arn,
 *     policy: amplifyAppMasterPolicyDocument.apply(amplifyAppMasterPolicyDocument => amplifyAppMasterPolicyDocument.json),
 * });
 * ```
 *
 * ## Import
 *
 * Amplify branch can be imported using `app_id` and `branch_name`, e.g.,
 *
 * ```sh
 *  $ pulumi import aws:amplify/branch:Branch master d2ypk4k47z8u6/master
 * ```
 */
class Branch extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["appId"] = state ? state.appId : undefined;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["associatedResources"] = state ? state.associatedResources : undefined;
            inputs["backendEnvironmentArn"] = state ? state.backendEnvironmentArn : undefined;
            inputs["basicAuthCredentials"] = state ? state.basicAuthCredentials : undefined;
            inputs["branchName"] = state ? state.branchName : undefined;
            inputs["customDomains"] = state ? state.customDomains : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["destinationBranch"] = state ? state.destinationBranch : undefined;
            inputs["displayName"] = state ? state.displayName : undefined;
            inputs["enableAutoBuild"] = state ? state.enableAutoBuild : undefined;
            inputs["enableBasicAuth"] = state ? state.enableBasicAuth : undefined;
            inputs["enableNotification"] = state ? state.enableNotification : undefined;
            inputs["enablePerformanceMode"] = state ? state.enablePerformanceMode : undefined;
            inputs["enablePullRequestPreview"] = state ? state.enablePullRequestPreview : undefined;
            inputs["environmentVariables"] = state ? state.environmentVariables : undefined;
            inputs["framework"] = state ? state.framework : undefined;
            inputs["pullRequestEnvironmentName"] = state ? state.pullRequestEnvironmentName : undefined;
            inputs["sourceBranch"] = state ? state.sourceBranch : undefined;
            inputs["stage"] = state ? state.stage : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["tagsAll"] = state ? state.tagsAll : undefined;
            inputs["ttl"] = state ? state.ttl : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.appId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'appId'");
            }
            if ((!args || args.branchName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'branchName'");
            }
            inputs["appId"] = args ? args.appId : undefined;
            inputs["backendEnvironmentArn"] = args ? args.backendEnvironmentArn : undefined;
            inputs["basicAuthCredentials"] = args ? args.basicAuthCredentials : undefined;
            inputs["branchName"] = args ? args.branchName : undefined;
            inputs["description"] = args ? args.description : undefined;
            inputs["displayName"] = args ? args.displayName : undefined;
            inputs["enableAutoBuild"] = args ? args.enableAutoBuild : undefined;
            inputs["enableBasicAuth"] = args ? args.enableBasicAuth : undefined;
            inputs["enableNotification"] = args ? args.enableNotification : undefined;
            inputs["enablePerformanceMode"] = args ? args.enablePerformanceMode : undefined;
            inputs["enablePullRequestPreview"] = args ? args.enablePullRequestPreview : undefined;
            inputs["environmentVariables"] = args ? args.environmentVariables : undefined;
            inputs["framework"] = args ? args.framework : undefined;
            inputs["pullRequestEnvironmentName"] = args ? args.pullRequestEnvironmentName : undefined;
            inputs["stage"] = args ? args.stage : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["ttl"] = args ? args.ttl : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["associatedResources"] = undefined /*out*/;
            inputs["customDomains"] = undefined /*out*/;
            inputs["destinationBranch"] = undefined /*out*/;
            inputs["sourceBranch"] = undefined /*out*/;
            inputs["tagsAll"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(Branch.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Branch resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Branch(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Branch.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Branch.__pulumiType;
    }
}
exports.Branch = Branch;
/** @internal */
Branch.__pulumiType = 'aws:amplify/branch:Branch';
