// @pulumi/aws/cognito/userPoolUICustomization.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserPoolUICustomization = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a Cognito User Pool UI Customization resource.
 *
 * > **Note:** To use this resource, the user pool must have a domain associated with it. For more information, see the Amazon Cognito Developer Guide on [Customizing the Built-in Sign-In and Sign-up Webpages](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-app-ui-customization.html).
 *
 * ## Example Usage
 * ### UI customization settings for a single client
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 * import * from "fs";
 *
 * const exampleUserPool = new aws.cognito.UserPool("exampleUserPool", {});
 * const exampleUserPoolDomain = new aws.cognito.UserPoolDomain("exampleUserPoolDomain", {
 *     domain: "example",
 *     userPoolId: exampleUserPool.id,
 * });
 * const exampleUserPoolClient = new aws.cognito.UserPoolClient("exampleUserPoolClient", {userPoolId: exampleUserPool.id});
 * const exampleUserPoolUICustomization = new aws.cognito.UserPoolUICustomization("exampleUserPoolUICustomization", {
 *     clientId: exampleUserPoolClient.id,
 *     css: ".label-customizable {font-weight: 400;}",
 *     imageFile: Buffer.from(fs.readFileSync("logo.png"), 'binary').toString('base64'),
 *     userPoolId: exampleUserPoolDomain.userPoolId,
 * });
 * ```
 * ### UI customization settings for all clients
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 * import * from "fs";
 *
 * const exampleUserPool = new aws.cognito.UserPool("exampleUserPool", {});
 * const exampleUserPoolDomain = new aws.cognito.UserPoolDomain("exampleUserPoolDomain", {
 *     domain: "example",
 *     userPoolId: exampleUserPool.id,
 * });
 * const exampleUserPoolUICustomization = new aws.cognito.UserPoolUICustomization("exampleUserPoolUICustomization", {
 *     css: ".label-customizable {font-weight: 400;}",
 *     imageFile: Buffer.from(fs.readFileSync("logo.png"), 'binary').toString('base64'),
 *     userPoolId: exampleUserPoolDomain.userPoolId,
 * });
 * ```
 *
 * ## Import
 *
 * Cognito User Pool UI Customizations can be imported using the `user_pool_id` and `client_id` separated by `,`, e.g.,
 *
 * ```sh
 *  $ pulumi import aws:cognito/userPoolUICustomization:UserPoolUICustomization example us-west-2_ZCTarbt5C,12bu4fuk3mlgqa2rtrujgp6egq
 * ```
 */
class UserPoolUICustomization extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["clientId"] = state ? state.clientId : undefined;
            inputs["creationDate"] = state ? state.creationDate : undefined;
            inputs["css"] = state ? state.css : undefined;
            inputs["cssVersion"] = state ? state.cssVersion : undefined;
            inputs["imageFile"] = state ? state.imageFile : undefined;
            inputs["imageUrl"] = state ? state.imageUrl : undefined;
            inputs["lastModifiedDate"] = state ? state.lastModifiedDate : undefined;
            inputs["userPoolId"] = state ? state.userPoolId : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.userPoolId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'userPoolId'");
            }
            inputs["clientId"] = args ? args.clientId : undefined;
            inputs["css"] = args ? args.css : undefined;
            inputs["imageFile"] = args ? args.imageFile : undefined;
            inputs["userPoolId"] = args ? args.userPoolId : undefined;
            inputs["creationDate"] = undefined /*out*/;
            inputs["cssVersion"] = undefined /*out*/;
            inputs["imageUrl"] = undefined /*out*/;
            inputs["lastModifiedDate"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(UserPoolUICustomization.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing UserPoolUICustomization resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new UserPoolUICustomization(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of UserPoolUICustomization.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === UserPoolUICustomization.__pulumiType;
    }
}
exports.UserPoolUICustomization = UserPoolUICustomization;
/** @internal */
UserPoolUICustomization.__pulumiType = 'aws:cognito/userPoolUICustomization:UserPoolUICustomization';
