// @pulumi/aws/cloudwatch/eventConnection.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventConnection = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides an EventBridge connection resource.
 *
 * > **Note:** EventBridge was formerly known as CloudWatch Events. The functionality is identical.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const test = new aws.cloudwatch.EventConnection("test", {
 *     authParameters: {
 *         apiKey: {
 *             key: "x-signature",
 *             value: "1234",
 *         },
 *     },
 *     authorizationType: "API_KEY",
 *     description: "A connection description",
 * });
 * ```
 * ### Basic Authorization
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const test = new aws.cloudwatch.EventConnection("test", {
 *     authParameters: {
 *         basic: {
 *             password: "Pass1234!",
 *             username: "user",
 *         },
 *     },
 *     authorizationType: "BASIC",
 *     description: "A connection description",
 * });
 * ```
 * ### OAuth Authorization
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const test = new aws.cloudwatch.EventConnection("test", {
 *     authParameters: {
 *         oauth: {
 *             authorizationEndpoint: "https://auth.url.com/endpoint",
 *             clientParameters: {
 *                 clientId: "1234567890",
 *                 clientSecret: "Pass1234!",
 *             },
 *             httpMethod: "GET",
 *             oauthHttpParameters: {
 *                 bodies: [{
 *                     isValueSecret: false,
 *                     key: "body-parameter-key",
 *                     value: "body-parameter-value",
 *                 }],
 *                 headers: [{
 *                     isValueSecret: false,
 *                     key: "header-parameter-key",
 *                     value: "header-parameter-value",
 *                 }],
 *                 queryStrings: [{
 *                     isValueSecret: false,
 *                     key: "query-string-parameter-key",
 *                     value: "query-string-parameter-value",
 *                 }],
 *             },
 *         },
 *     },
 *     authorizationType: "OAUTH_CLIENT_CREDENTIALS",
 *     description: "A connection description",
 * });
 * ```
 * ### Invocation Http Parameters
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const test = new aws.cloudwatch.EventConnection("test", {
 *     authParameters: {
 *         basic: {
 *             password: "Pass1234!",
 *             username: "user",
 *         },
 *         invocationHttpParameters: {
 *             bodies: [
 *                 {
 *                     isValueSecret: false,
 *                     key: "body-parameter-key",
 *                     value: "body-parameter-value",
 *                 },
 *                 {
 *                     isValueSecret: true,
 *                     key: "body-parameter-key2",
 *                     value: "body-parameter-value2",
 *                 },
 *             ],
 *             headers: [{
 *                 isValueSecret: false,
 *                 key: "header-parameter-key",
 *                 value: "header-parameter-value",
 *             }],
 *             queryStrings: [{
 *                 isValueSecret: false,
 *                 key: "query-string-parameter-key",
 *                 value: "query-string-parameter-value",
 *             }],
 *         },
 *     },
 *     authorizationType: "BASIC",
 *     description: "A connection description",
 * });
 * ```
 *
 * ## Import
 *
 * EventBridge Connection can be imported using the `name`, e.g., console
 *
 * ```sh
 *  $ pulumi import aws:cloudwatch/eventConnection:EventConnection test ngrok-connection
 * ```
 */
class EventConnection extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["authParameters"] = state ? state.authParameters : undefined;
            inputs["authorizationType"] = state ? state.authorizationType : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["secretArn"] = state ? state.secretArn : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.authParameters === undefined) && !opts.urn) {
                throw new Error("Missing required property 'authParameters'");
            }
            if ((!args || args.authorizationType === undefined) && !opts.urn) {
                throw new Error("Missing required property 'authorizationType'");
            }
            inputs["authParameters"] = args ? args.authParameters : undefined;
            inputs["authorizationType"] = args ? args.authorizationType : undefined;
            inputs["description"] = args ? args.description : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["secretArn"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(EventConnection.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing EventConnection resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new EventConnection(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of EventConnection.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === EventConnection.__pulumiType;
    }
}
exports.EventConnection = EventConnection;
/** @internal */
EventConnection.__pulumiType = 'aws:cloudwatch/eventConnection:EventConnection';
