// @pulumi/aws/synthetics/canary.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Canary = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a Synthetics Canary resource.
 *
 * > **NOTE:** When you create a canary, AWS creates supporting implicit resources. See the Amazon CloudWatch Synthetics documentation on [DeleteCanary](https://docs.aws.amazon.com/AmazonSynthetics/latest/APIReference/API_DeleteCanary.html) for a full list. Neither AWS nor this provider deletes these implicit resources automatically when the canary is deleted. Before deleting a canary, ensure you have all the information about the canary that you need to delete the implicit resources using the AWS Console, or AWS CLI.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const some = new aws.synthetics.Canary("some", {
 *     artifactS3Location: "s3://some-bucket/",
 *     executionRoleArn: "some-role",
 *     handler: "exports.handler",
 *     runtimeVersion: "syn-1.0",
 *     schedule: {
 *         expression: "rate(0 minute)",
 *     },
 *     zipFile: "test-fixtures/lambdatest.zip",
 * });
 * ```
 *
 * ## Import
 *
 * Synthetics Canaries can be imported using the `name`, e.g.,
 *
 * ```sh
 *  $ pulumi import aws:synthetics/canary:Canary some some-canary
 * ```
 */
class Canary extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["artifactS3Location"] = state ? state.artifactS3Location : undefined;
            inputs["engineArn"] = state ? state.engineArn : undefined;
            inputs["executionRoleArn"] = state ? state.executionRoleArn : undefined;
            inputs["failureRetentionPeriod"] = state ? state.failureRetentionPeriod : undefined;
            inputs["handler"] = state ? state.handler : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["runConfig"] = state ? state.runConfig : undefined;
            inputs["runtimeVersion"] = state ? state.runtimeVersion : undefined;
            inputs["s3Bucket"] = state ? state.s3Bucket : undefined;
            inputs["s3Key"] = state ? state.s3Key : undefined;
            inputs["s3Version"] = state ? state.s3Version : undefined;
            inputs["schedule"] = state ? state.schedule : undefined;
            inputs["sourceLocationArn"] = state ? state.sourceLocationArn : undefined;
            inputs["startCanary"] = state ? state.startCanary : undefined;
            inputs["status"] = state ? state.status : undefined;
            inputs["successRetentionPeriod"] = state ? state.successRetentionPeriod : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["tagsAll"] = state ? state.tagsAll : undefined;
            inputs["timelines"] = state ? state.timelines : undefined;
            inputs["vpcConfig"] = state ? state.vpcConfig : undefined;
            inputs["zipFile"] = state ? state.zipFile : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.artifactS3Location === undefined) && !opts.urn) {
                throw new Error("Missing required property 'artifactS3Location'");
            }
            if ((!args || args.executionRoleArn === undefined) && !opts.urn) {
                throw new Error("Missing required property 'executionRoleArn'");
            }
            if ((!args || args.handler === undefined) && !opts.urn) {
                throw new Error("Missing required property 'handler'");
            }
            if ((!args || args.runtimeVersion === undefined) && !opts.urn) {
                throw new Error("Missing required property 'runtimeVersion'");
            }
            if ((!args || args.schedule === undefined) && !opts.urn) {
                throw new Error("Missing required property 'schedule'");
            }
            inputs["artifactS3Location"] = args ? args.artifactS3Location : undefined;
            inputs["executionRoleArn"] = args ? args.executionRoleArn : undefined;
            inputs["failureRetentionPeriod"] = args ? args.failureRetentionPeriod : undefined;
            inputs["handler"] = args ? args.handler : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["runConfig"] = args ? args.runConfig : undefined;
            inputs["runtimeVersion"] = args ? args.runtimeVersion : undefined;
            inputs["s3Bucket"] = args ? args.s3Bucket : undefined;
            inputs["s3Key"] = args ? args.s3Key : undefined;
            inputs["s3Version"] = args ? args.s3Version : undefined;
            inputs["schedule"] = args ? args.schedule : undefined;
            inputs["startCanary"] = args ? args.startCanary : undefined;
            inputs["successRetentionPeriod"] = args ? args.successRetentionPeriod : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["vpcConfig"] = args ? args.vpcConfig : undefined;
            inputs["zipFile"] = args ? args.zipFile : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["engineArn"] = undefined /*out*/;
            inputs["sourceLocationArn"] = undefined /*out*/;
            inputs["status"] = undefined /*out*/;
            inputs["tagsAll"] = undefined /*out*/;
            inputs["timelines"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(Canary.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Canary resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Canary(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Canary.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Canary.__pulumiType;
    }
}
exports.Canary = Canary;
/** @internal */
Canary.__pulumiType = 'aws:synthetics/canary:Canary';
