// @pulumi/aws/cloudfront/cachePolicy.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.CachePolicy = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * ## Example Usage
 *
 * The following example below creates a CloudFront cache policy.
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.cloudfront.CachePolicy("example", {
 *     comment: "test comment",
 *     defaultTtl: 50,
 *     maxTtl: 100,
 *     minTtl: 1,
 *     parametersInCacheKeyAndForwardedToOrigin: {
 *         cookiesConfig: {
 *             cookieBehavior: "whitelist",
 *             cookies: {
 *                 items: ["example"],
 *             },
 *         },
 *         headersConfig: {
 *             headerBehavior: "whitelist",
 *             headers: {
 *                 items: ["example"],
 *             },
 *         },
 *         queryStringsConfig: {
 *             queryStringBehavior: "whitelist",
 *             queryStrings: {
 *                 items: ["example"],
 *             },
 *         },
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * Cloudfront Cache Policies can be imported using the `id`, e.g.,
 *
 * ```sh
 *  $ pulumi import aws:cloudfront/cachePolicy:CachePolicy policy 658327ea-f89d-4fab-a63d-7e88639e58f6
 * ```
 */
class CachePolicy extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["comment"] = state ? state.comment : undefined;
            inputs["defaultTtl"] = state ? state.defaultTtl : undefined;
            inputs["etag"] = state ? state.etag : undefined;
            inputs["maxTtl"] = state ? state.maxTtl : undefined;
            inputs["minTtl"] = state ? state.minTtl : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["parametersInCacheKeyAndForwardedToOrigin"] = state ? state.parametersInCacheKeyAndForwardedToOrigin : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.parametersInCacheKeyAndForwardedToOrigin === undefined) && !opts.urn) {
                throw new Error("Missing required property 'parametersInCacheKeyAndForwardedToOrigin'");
            }
            inputs["comment"] = args ? args.comment : undefined;
            inputs["defaultTtl"] = args ? args.defaultTtl : undefined;
            inputs["maxTtl"] = args ? args.maxTtl : undefined;
            inputs["minTtl"] = args ? args.minTtl : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["parametersInCacheKeyAndForwardedToOrigin"] = args ? args.parametersInCacheKeyAndForwardedToOrigin : undefined;
            inputs["etag"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(CachePolicy.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing CachePolicy resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new CachePolicy(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of CachePolicy.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === CachePolicy.__pulumiType;
    }
}
exports.CachePolicy = CachePolicy;
/** @internal */
CachePolicy.__pulumiType = 'aws:cloudfront/cachePolicy:CachePolicy';
